Cleanup: Move k8s API calls to internal/util/k8s

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

- cleanup: move k8s API calls to interna/util/k8s 

  Move k8s API calls for VolumeAttachment and PersistentVolume
  from internal/rbd/rbd_healer to internal/util/k8s/
  {volumeattachments.go,persistentvolumes.go}.

- cleanup: move k8s ConfigMap API calls to internal/util/k8s 

  Move k8s API calls for ConfigMap from internal/kms to new
  internal/util/k8s/configmap.go file.

- cleanup: move k8s ServiceAccount API calls to internal/util/k8s 

  Move k8s API calls for ServiceAccount from internal/kms to new
  internal/util/k8s/serviceaccounts.go file.


**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
cleanup: move k8s API calls to interna/util/k8s

Move k8s API calls for VolumeAttachment and PersistentVolume
from internal/rbd/rbd_healer to internal/util/k8s/
{volumeattachments.go,persistentvolumes.go}.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cleanup: move k8s ConfigMap API calls to internal/util/k8s

Move k8s API calls for ConfigMap from internal/kms to new
internal/util/k8s/configmap.go file.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cleanup: move k8s ServiceAccount API calls to internal/util/k8s

Move k8s API calls for ServiceAccount from internal/kms to new
internal/util/k8s/serviceaccounts.go file.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.33

Rakshith-R reviewed (commented):
@iPraveenParihar 
maybe you want to improve https://github.com/ceph/ceph-csi/blob/devel/internal/util/k8s/client.go#L32 to only create client once and reuse it.

let us know if you want to do it in this pr or another one. 

Everything else looks good to me.

nixpanic commented on internal/rbd/rbd_healer.go:
```diff
@@ -75,12 +72,12 @@ func getSecret(ns, name string) (map[string]string, error) {
 	targetPath := filepath.Join(stagingPath, pv.Spec.CSI.Driver, hex.EncodeToString(unique[:]), "globalmount")
 
-	major, minor, err := kubeclient.GetServerVersion(c)
+	major, minor, err := k8s.GetServerVersion()
```
This isn't even needed anymore. The check is for a Kubernetes version that we don't support since a while.

nixpanic commented on internal/rbd/rbd_healer.go:
```diff
@@ -95,11 +92,11 @@ func formatStagingTargetPath(c *k8s.Clientset, pv *v1.PersistentVolume, stagingP
 }
 
-func callNodeStageVolume(ns *NodeServer, c *k8s.Clientset, pv *v1.PersistentVolume, stagingPath string) error {
+func callNodeStageVolume(ns *NodeServer, pv *v1.PersistentVolume, stagingPath string) error {
```
Any function that gets the `NodeServer` as parameter is a good candidate to become a function of the `NodeServer struct`.

This is something that you may want to cleanup in a follow-up PR (or separate commit).

nixpanic commented on internal/util/k8s/persistentvolumes.go:
```diff
@@ -0,0 +1,45 @@
+		return nil, fmt.Errorf("failed to connect to Kubernetes: %w", err)
+	}
+
+	pv, err := client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
```
Eventually all `context.TODO()` should be replaced by the actual context for the gRPC call. Not all interactions with Kubernetes may come through a gRPC call, and those should create their own context (as high in the stack as possible).

Just an other cleanup task for the future :smile:

iPraveenParihar replied:
Let's do that in a different PR 😄. I'll create a tracker for it.

iPraveenParihar replied:
tracker: https://github.com/ceph/ceph-csi/issues/5668

nixpanic commented on internal/util/k8s/configmap.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// GetConfigMap retrieves a Kubernetes ConfigMap by its name and namespace.
+func GetConfigMap(configMapName, configMapNamespace string, ignoreNotFound bool) (*corev1.ConfigMap, error) {
```
nit: To me it feels more intuitive to have the namespace as first parameter. Most `kubectl` commands will use the format of `namespace/objectName` too.

iPraveenParihar committed:
cleanup: remove k8s version check in staging target path formatting

Since ceph-csi no longer supports k8s versions older than 1.24,
this commit removes the version check for 1.24 in formatStagingTargetPath.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
rbd: convert functions to NodeServer struct  methods

This change eliminates the need to pass the NodeServer
instance as a parameter and follows Go best practices.

Signed-off-by: Praveen M <m.praveen@ibm.com>

Madhu-1 commented on internal/util/k8s/serviceaccounts.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+// CreateServiceAccountToken creates a token for the given ServiceAccount.
+func CreateServiceAccountToken(saNamespace, saName string) (*authenticationv1.TokenRequest, error) {
```
```suggestion
func CreateServiceAccountToken(namespace, name string) (*authenticationv1.TokenRequest, error) {
```

Madhu-1 commented on internal/util/k8s/serviceaccounts.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// GetServiceAccount retrieves the ServiceAccount object for the given name and namespace.
+func GetServiceAccount(saNamespace, saName string) (*corev1.ServiceAccount, error) {
```
```suggestion
func GetServiceAccount(namespace, name string) (*corev1.ServiceAccount, error) {
```

Madhu-1 commented on internal/util/k8s/volumeattachments.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// GetVolumeAttachment returns the VolumeAttachment object for the given vaName.
+func GetVolumeAttachment(vaName string, ignoreNotFound bool) (*storagev1.VolumeAttachment, error) {
```
```suggestion
func GetVolumeAttachment(name string, ignoreNotFound bool) (*storagev1.VolumeAttachment, error) {
```

Madhu-1 replied:
Passing `ignoreNotFound` should not be part of the input param. you can expose a function to check the error type

iPraveenParihar replied:
I was thinking of avoiding k8s errors check from caller.

Madhu-1 replied:
you can write an wrapper in utils if you want to avoid including the k8s api at the caller

Madhu-1 commented on internal/util/k8s/persistentvolumes.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// GetPersistentVolume returns the PersistentVolume object for the given pvName.
+func GetPersistentVolume(pvName string, ignoreNotFound bool) (*corev1.PersistentVolume, error) {
```
```suggestion
func GetPersistentVolume(name string, ignoreNotFound bool) (*corev1.PersistentVolume, error) {
```

same here please dont pass `ignoreNotFound` as an input

Madhu-1 commented on internal/util/k8s/configmap.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// GetConfigMap retrieves a Kubernetes ConfigMap by its name and namespace.
+func GetConfigMap(configMapNamespace, configMapName string, ignoreNotFound bool) (*corev1.ConfigMap, error) {
```
```suggestion
func GetConfigMap(namespace, name string, ignoreNotFound bool) (*corev1.ConfigMap, error) {
```

same here as well, dont pass `ignoreNotFound` as input

Madhu-1 reviewed (changes requested):
please open two trackers for below
* Reuse of client
* Passing the context

Madhu-1 commented on internal/util/k8s/client.go:
```diff
@@ -66,3 +68,13 @@ func RunsOnKubernetes() bool {
+	}
+
+	return err
+}
```
new line is required at the end of the file.

Madhu-1 reviewed (commented):
LGTM. small nit. i assume we already have tracker for having single K8s client if not lets open 1

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
Stopped most of the CI jobs, they are running at #5733 already too (and Mergify only cares about those results).
